add TCP StatsD listener support

jwfang opened the pull request:
support TCP StatsD, on port 9125

SuperQ commented on main.go:
```diff
@@ -32,13 +32,14 @@ func init() {
+	listenAddress          = flag.String("web.listen-address", ":9102", "The address on which to expose the web interface and generated Prometheus metrics.")
+	metricsEndpoint        = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics.")
+	statsdListenAddress    = flag.String("statsd.listen-address", ":9125", "The UDP address on which to receive statsd metric lines.")
+	statsdTCPListenAddress = flag.String("statsd.tcp-listen-address", ":9126", "The TCP address on which to receive statsd metric lines.")
```
I'd rather keep this one flag, so that it listens on the same TCP/UDP port.  Also it looks like you have a typo, it's listening on `9126` here.

jwfang replied:
good catch, i fixed it.

i will make this optional.

SuperQ commented:
Instead of making this UDP or TCP, I would just have it listen for both.

jwfang commented:
sorry, misunderstand you.  

really make more sense to listen on both.

SuperQ commented:
No problem, thanks for working on this feature! 😃

SuperQ commented on main.go:
```diff
@@ -34,7 +34,9 @@ func init() {
-	statsdListenAddress = flag.String("statsd.listen-address", ":9125", "The UDP address on which to receive statsd metric lines.")
+	statsdListenAddress = flag.String("statsd.listen-address", ":9125", "The UDP/TCP address on which to receive statsd metric lines.")
+	statsdListenUDP     = flag.Bool("statsd.listen-udp", true, "Whether to receive UDP statsd metrics.")
+	statsdListenTCP     = flag.Bool("statsd.listen-tcp", false, "Whether to receive TCP statsd metrics.")
```
This can be `true` by default.

SuperQ commented on exporter.go:
```diff
@@ -351,105 +338,159 @@ func parseDogStatsDTagsToLabels(component string) map[string]string {
+	return
+}
+
+type StatsDListener struct {
```
Can we make this `StatsDUDPListener` for consistency?

jwfang replied:
i thought UDP is primary and TCP is secondary originally.

done rename.

SuperQ commented:
Ping @juliusv Any more comments?

jacksontj commented on exporter.go:
```diff
@@ -351,105 +338,159 @@ func parseDogStatsDTagsToLabels(component string) map[string]string {
+		if err != nil {
+			log.Fatalf("AcceptTCP failed: %v", err)
+		}
+		go l.handleConn(c, e)
```
Before calling handleconn (or immediately in it) -- we should probably call `SetReadBuffer` -- to be consistent with UDP

jwfang replied:
UDP needs a 64K buffer, since it will recv all data in one packet.
but for TCP, i guess the underlying bufio is manage the buffer for us.

jacksontj commented on exporter.go:
```diff
@@ -351,105 +338,159 @@ func parseDogStatsDTagsToLabels(component string) map[string]string {
 	e <- events
 }
+
+type StatsDTCPListener struct {
```
We don't *need* a separate listener -- assuming we have a type switch. It seems like it'd be *less* code-- but this definitely isn't wrong ;)

jwfang commented on exporter.go:
```diff
@@ -351,105 +338,159 @@ func parseDogStatsDTagsToLabels(component string) map[string]string {
+
+	r := bufio.NewReader(c)
+	for {
+		line, isPrefix, err := r.ReadLine()
```
ReadBytes will read till *delimiter* or *eof*, so i think ReadLine is more safe for bogus input with really long line, or no line delimiter at all. :)

jacksontj replied:
Fair enough, LGTM then :) Lets merge it!

SuperQ commented:
@juliusv care to take one last look?

grobie commented on exporter.go:
```diff
@@ -351,105 +338,159 @@ func parseDogStatsDTagsToLabels(component string) map[string]string {
-		if line == "" {
-			continue
-		}
+func lineToEvents(line string) (events Events) {
```
Please remove the named return variable here as it adds stutter to the function signature and use explicit returns below. It's not a go idiom to use naked returns in such long functions.

> Named returned values should only be used on public funcs and methods
when it contributes to the documentation.

https://go-review.googlesource.com/c/20024/

grobie commented on exporter.go:
```diff
@@ -351,105 +338,159 @@ func parseDogStatsDTagsToLabels(component string) map[string]string {
+		line, isPrefix, err := r.ReadLine()
+		if err != nil {
+			if err != io.EOF {
+				log.Errorf("Read %s failed: %v", c.RemoteAddr(), err)
```
It'd be cool to have metrics for these errors so that operators can set up alerts on misbehaving clients.

grobie commented on main.go:
```diff
@@ -122,34 +138,53 @@ func main() {
 	}
 
+	if !*statsdListenUDP && !*statsdListenTCP {
+		log.Fatalln("At least one of UDP/TCP listeners should be specified.")
```
Sounds like a `must` then.

grobie commented on main.go:
```diff
@@ -34,7 +34,9 @@ func init() {
 	metricsEndpoint     = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics.")
-	statsdListenAddress = flag.String("statsd.listen-address", ":9125", "The UDP address on which to receive statsd metric lines.")
+	statsdListenAddress = flag.String("statsd.listen-address", ":9125", "The UDP/TCP address on which to receive statsd metric lines.")
+	statsdListenUDP     = flag.Bool("statsd.listen-udp", true, "Whether to receive UDP statsd metrics.")
```
Instead of boolean flags and introducing a dependency between flags, it'd be simpler to just have to flags with the addresses for each protocol, like `-statsd.listen-address-udp=":9125"` `-statsd.listen-address-tcp=":9125"`. An empty value signals disables the server on that protocol.

jwfang replied:
doing it like what you mentioned have two conerns for me:
1. a bit uncomfortable for setting something to "" to disable it; and since we're default both UDP/TCP to on, use have to explicitly disable it;
2. the flags changes will break the user interface, from *statsdListenAddress* to *statsdListenUDP*. 

i have no strong feeling for 1; and for 2, if that's OK, i will do the flags changes.

grobie commented on main.go:
```diff
@@ -122,34 +138,53 @@ func main() {
 	log.Infoln("Starting StatsD -> Prometheus Exporter", version.Info())
 	log.Infoln("Build context", version.BuildContext())
-	log.Infoln("Accepting StatsD Traffic on", *statsdListenAddress)
+	log.Infof("Accepting StatsD Traffic on %s, UDP %v, TCP %v", *statsdListenAddress, *statsdListenUDP, *statsdListenTCP)
```
That's not longer printing the address it's listening on, but `true`/`false`, but would be fixed with the comment above.

jwfang commented:
@grobie thanks for your reivew, addressed most of your comments except for the test for TCPListener.

can't find some easy way to mock the TCPConnection. maybe we can make a 
`func (l *StatsDTCPListener) handleConn(r *io.Reader, e chan<- Events) `
 but that made the stats/logging a bit awkward. :(

grobie commented:
> can't find some easy way to mock the TCPConnection

I'd say, then don't mock for now and just start the server, send it a statsd line via TCP and check that it was ingested properly (by looking at the exported metrics).

jwfang commented:
@grobie add test for TCP listener in the same place as UDP except benchmarks.

grobie commented:
Thanks a lot for your contribution @jwfang!
